src/libfsm/vm/vm.c

```diff
@@ -473,3 +473,18 @@ dfavm_compile_vm(const struct dfavm_assembler_ir *a, struct fsm_vm_compile_opts
 	return NULL;
 }
 
+void
+dfavm_free_vm(struct fsm_dfavm *vm)
+{
+	if (vm->version_major == DFAVM_VARENC_MAJOR && vm->version_minor == DFAVM_VARENC_MINOR) {
+		dfavm_v1_finalize(&vm->u.v1);
+	} else if (vm->version_major == DFAVM_FIXEDENC_MAJOR && vm->version_minor == DFAVM_FIXEDENC_MINOR) {
+		dfavm_v2_finalize(&vm->u.v2);
+	} else {
+		/* invalid VM version! */
+		/* XXX - should we do something? */
```

Comment:
Probably assert on it. It's not part of the calling contract to deal with invalid things passed here.

I would allow for `vm` to be NULL though, in the spirit of `free()`.

Reply:
I think I've fixed these in 01c3fa9 and 67f3978